Extend `tokens` function with additional tokenizers

ahmadov opened the pull request:
Extend the existing `tokens` function with multiple tokenizers support. The supported tokenizers are `default`, `noop`, `ngram`. Deprecated the existing `ngrams` function in favor of `tokens` function with the `ngram` tokenizer.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The `tokens` function was extended to accept an additional "tokenizer" argument plus further tokenizer-specific arguments.

ahmadov committed:
Add ChineseTokenizer and tokenizeChinese function

The Chinese tokenizer is supported via `cppjieba` [[1]].

[1]: https://github.com/yanyiwu/cppjieba

rschu1ze commented on ci/jobs/fast_test.py:
```diff
@@ -55,6 +55,8 @@ def clone_submodules():
         "contrib/incbin",
         "contrib/yaml-cpp",
         "contrib/corrosion",
+        "contrib/limonp",
```
The list defines which submodules are compiled into ClickHouse for the "Fasttests" CI check (which is supposed to be a smoke test). Both new submodules are not strictly necessary. Let's remove them from here.

rschu1ze commented on contrib/cppjieba-cmake/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+# limonp is dependency of cppjieba.
```
Regarding my other comment about Fasttest, we should make the build of cppjieba dependent on whether CMake variable `ENABLE_LIBRARIES` is set.

Stealing some bits from contrib/sqids-cpp-cmake/CMakeLists.txt (the CMake build description of another non-essential third-party library), this is how that can be done:

```cmake
option(ENABLE_SQIDS "Enable sqids support" ${ENABLE_LIBRARIES})
if ((NOT ENABLE_SQIDS))
    message (STATUS "Not using sqids")
    return()
endif()
```

If we go that route, we will need to
- ifdef all code which uses cppjieba (see `USE_SQIDS` in src/Functions/sqid.cpp for an example), and
- exclude SQL tests for cppjieba from FastTest (simply add `-- Tags: no-fasttest` at the beginning of the SQL test)

rschu1ze commented on src/Common/Tokenizer/ChineseTokenizer.cpp:
```diff
@@ -0,0 +1,32 @@
+
+ChineseTokenizer & ChineseTokenizer::instance()
+{
+    static ChineseTokenizer tokenizer{};
```
(`{}` isn't needed, afaik)

rschu1ze commented on src/Common/Tokenizer/ChineseTokenizer.h:
```diff
@@ -0,0 +1,27 @@
+namespace DB
+{
+
+class ChineseTokenizer
```
Add a class-level comment that this class is a wrapper around libcppjieba?

rschu1ze commented on src/Common/Tokenizer/ChineseTokenizer.cpp:
```diff
@@ -0,0 +1,32 @@
+}
+
+ChineseTokenizer::ChineseTokenizer()
+    : jieba_instance{std::make_unique<cppjieba::Jieba>()}
```
(cosmetic: uniform initialization isn't very popular across the code base, let's use `()`)

rschu1ze commented on src/Common/Tokenizer/ChineseTokenizer.cpp:
```diff
@@ -0,0 +1,32 @@
+
+std::vector<std::string> ChineseTokenizer::tokenize(const std::string & str)
+{
+#ifdef DEBUG_OR_SANITIZER_BUILD
```
The singleton code is quite simple, let's remove l. 19-21 ... it is too obvious that the assert can never be hit.

rschu1ze commented on src/Common/Tokenizer/ChineseTokenizer.cpp:
```diff
@@ -0,0 +1,32 @@
+    assert(jieba_instance != nullptr);
+#endif
+    std::vector<std::string> tokens{};
+    jieba_instance->CutForSearch(str, tokens);
```
Hmm, as far as I remembered Jieba accepts a "fine-granular" vs. "coarse-granular" mode parameter but I can't find it anywhere around here.

ahmadov replied:
There is no parameter for "fine-granular" and "coarse-granular" mode parameters. There are two different methods for that purpose: `jieba_instance->CutForSearch` and `jieba_instance->Cut`.

rschu1ze replied:
Right ... I guess my question is if we could distinguish between fine- and coarse-granular at the SQL function level, i.e. add another parameter for users to specify either mode (or go with the default), and then internally call the right Jieba function.

ahmadov replied:
I removed the `tokenizeChinese` method and added a generic `tokenize` method to support multiple tokenizers: `default`, `ngram`, `noop` and `chinese` (enabled when ENABLE_CPPJIEBA is set to ON). I also added `fine-grained` and `coarse-grained` modes to the ChineseTokenizer. You can see the examples

rschu1ze commented on src/Functions/tokenizeChinese.cpp:
```diff
@@ -0,0 +1,103 @@
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+    bool isVariadic() const override { return false; }
```
isVariadic = false is the default anyways (see `IFunction`). Let's remove l. 35.

rschu1ze commented on src/Functions/tokenizeChinese.cpp:
```diff
@@ -0,0 +1,103 @@
+
+    size_t getNumberOfArguments() const override { return 1; }
+    bool isVariadic() const override { return false; }
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return ColumnNumbers{}; }
```
Same for l. 36.

rschu1ze commented on src/Functions/tokenizeChinese.cpp:
```diff
@@ -0,0 +1,103 @@
+    bool isVariadic() const override { return false; }
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return ColumnNumbers{}; }
+
+    bool useDefaultImplementationForNulls() const override { return true; }
```
... and l. 38-40.

rschu1ze commented on src/Functions/tokenizeChinese.cpp:
```diff
@@ -0,0 +1,103 @@
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        auto input_argument_type = WhichDataType(arguments[0].type);
```
For argument validation, let's use `FunctionArgumentDescriptors` from src/Functions/FunctionHelpers.h - it is more descriptive than manual validation. Please see src/Functions/makeDate.cpp for an example how to use it.

rschu1ze commented on src/Functions/tokenizeChinese.cpp:
```diff
@@ -0,0 +1,103 @@
+
+REGISTER_FUNCTION(TokenizerChinese)
+{
+    factory.registerFunction<FunctionTokenizeChinese>();
```
Needs `FunctionDocumentation` (please grep for this term in src/Functions/).

rschu1ze commented on src/Functions/tokenizeChinese.cpp:
```diff
@@ -0,0 +1,103 @@
+private:
+    template <typename StringColumnType>
+    void executeImpl(
+        StringColumnType & input_data_column,
```
\+ `const`

(cosmetic: people here like left-to-right hierarchical naming, i.e. `column_input` (or `col_input`) and `col_input_offsets` instead of `input_data_column` and `input_column_offsets`).

rschu1ze commented on src/Functions/tokenizeChinese.cpp:
```diff
@@ -0,0 +1,103 @@
+
+        for (size_t i = 0; i < input_rows_count; ++i)
+        {
+            const auto data = input_data_column.getDataAt(i);
```
Minro: I know `auto` is convenient but if it is not immediatly clear what the lhs type is (like in `auto x = static_cast<Something>(...)`) or if usage isn't idiomatic (like in `for (auto thing : things)`), then the explicit type is preferred.

rschu1ze commented on src/Functions/tokenizeChinese.cpp:
```diff
@@ -0,0 +1,103 @@
+        for (size_t i = 0; i < input_rows_count; ++i)
+        {
+            const auto data = input_data_column.getDataAt(i);
+            const auto words = ChineseTokenizer::instance().tokenize(data.data);
```
Consider pulling `words` out of the for loop to re-use it between iterations and save heap allocations (potentially). Instead call .clear() at the end of each iteration.

rschu1ze replied:
Oh, another thing ... could we implement a class `ChineseTokenExtractor` in src/Interpreters/ITokenExtractor.h which calls Jieba, and here only use this new class? That is basically how src/Functions/tokenExtractors.cpp is implemented.

rschu1ze commented on tests/queries/0_stateless/03403_function_tokenizeChinese.sql:
```diff
@@ -0,0 +1,47 @@
+SELECT tokenizeChinese('小明硕士毕业于中国科学院计算所，后在日本京都大学深造');
+SELECT length(tokenizeChinese('小明硕士毕业于中国科学院计算所，后在日本京都大学深造'));
+
+SELECT 'With materialized values.';
```
`materialized` --> `non-const`

rschu1ze commented on tests/queries/0_stateless/03403_function_tokenizeChinese.sql:
```diff
@@ -0,0 +1,47 @@
+SELECT 'With constant values.';
+SELECT tokenizeChinese('他来到了网易杭研大厦');
```
Let's rewrite to this form to avoid some copypasta.

```sql
WITH `他来到了网易杭研大厦` AS text
SELECT tokenizeChinese(text) AS tokenized, length(tokenized) as length;
```

rschu1ze commented on tests/queries/0_stateless/03403_function_tokenizeChinese.sql:
```diff
@@ -0,0 +1,47 @@
+SELECT 'With constant values.';
+SELECT tokenizeChinese('他来到了网易杭研大厦');
+SELECT length(tokenizeChinese('他来到了网易杭研大厦'));
```
@hanfei1991 Are the tokenized results okay or garbage?

hanfei1991 replied:
LGTM, it just copy/paste some tests from "jieba" repo :)

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/tokenizer-cn

ahmadov committed:
Remove `tokenizeChinese` and add a generic `tokenize` function

ahmadov committed:
Add 'fine-grained' and 'coarse-grained' modes to ChineseTokenExtractor

ahmadov committed:
include more examples in documentation

- includes style fix

ahmadov committed:
extend tokens function with additional additional tokenizers

- avoid duplication for validating arguments
- Make `tokenizer` a drop-in replacement of `tokens`
- Add examples into documentation
- Split functions into different C++ files for better maintaining
- Deprecate `ngrams` function and document it

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into ahmadov/tokenizer-cn

rschu1ze committed:
Some fixups

rschu1ze commented:
@ahmadov I went over the PR and made a couple of changes for simplicity and constency with the rest of the codebase. Feel free to take a look.

(By coincidence, you made a force-push ca. an hour after I started fixing some stuff. As force-pushes rewrite the Git history, I wasn't able to understand the diff of what you changed - I probably (likely) overrode your latest changes, sorry. I could have gone back to the new HEAD but that would have been more work. As a general guideline, let's not force-push after someone is set as assignee.)

rschu1ze committed:
Clean up CMakeLists.txt

rschu1ze commented on .gitmodules:
```diff
@@ -272,6 +272,12 @@
 [submodule "contrib/ulid-c"]
 	path = contrib/ulid-c
 	url = https://github.com/ClickHouse/ulid-c.git
+[submodule "contrib/limonp"]
```
It's tempting to move new entries to the end of a list but it can also create useless merge conflicts with other PRs that do the same. To reduce that risk, I often choose a random position.

rschu1ze commented on contrib/CMakeLists.txt:
```diff
@@ -166,6 +166,12 @@ add_contrib (datasketches-cpp-cmake datasketches-cpp)
 
+option(ENABLE_CPPJIEBA "Enable cppjieba support" ${ENABLE_LIBRARIES})
+if (ENABLE_CPPJIEBA)
+    add_contrib (limonp-cmake limonp)
```
To make things more explicit, I split limonp into its own directory.

rschu1ze commented on docs/en/sql-reference/functions/splitting-merging-functions.md:
````diff
@@ -403,3 +413,31 @@ Result:
 └───────────────────────────────────┘
 ```
+
+Using the ngram tokenizer with ngram length 3:
````
Removed some of the examples to not overwhelm the reader (we should test all combinations in SQL tests but its not necessary to document all combinations in end user docs)

rschu1ze commented on src/Functions/tokens.cpp:
```diff
@@ -0,0 +1,166 @@
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors mandatory_args{
```
Simplified this method a bit.

rschu1ze commented on src/Functions/tokens.cpp:
```diff
@@ -0,0 +1,166 @@
+        if (input_rows_count == 0)
+            return ColumnArray::create(std::move(col_result), std::move(col_offsets));
+
+        std::unique_ptr<ITokenExtractor> token_extractor;
```
There was too much lambda magic here. I threw it out and replaced it by something more braindead.

rschu1ze commented on src/Functions/tokens.cpp:
```diff
@@ -0,0 +1,166 @@
+            token_extractor = std::make_unique<NoOpTokenExtractor>();
+        else if (tokenizer_arg == NgramTokenExtractor::getExternalName())
+        {
+            auto ngrams = arguments.size() < 3 ? 3
```
Minor change: Afaik, the 3rd ngrams argument was mandatory before if `ngrams` was given as tokenizer. Now its optional.

rschu1ze commented on src/Functions/tokens.cpp:
```diff
@@ -0,0 +1,166 @@
+        auto col_result = ColumnString::create();
+        auto col_offsets = ColumnArray::ColumnOffsets::create();
+
+        if (input_rows_count == 0)
```
That shortcut is important. Otherwise below code may read from empty columns and sanitizer builds will complain.

rschu1ze commented on src/Interpreters/ChineseTokenizer.cpp:
```diff
@@ -0,0 +1,38 @@
+std::vector<std::string> ChineseTokenizer::tokenize(const std::string & str, ChineseTokenizationGranularity granularity)
+{
+    std::vector<std::string> tokens;
+    switch (granularity)
```
Note: enums are ideally always iterated using switch statements. That way, the compiler will complain if someone adds a new enum value but forgets to add a new path. Such things will go unnoticed for if/else statements.

rschu1ze commented on src/Interpreters/ChineseTokenizer.h:
```diff
@@ -0,0 +1,41 @@
+
+#include "config.h"
+
+#if USE_CPPJIEBA
```
Added `USE_CPPJIEBA` all over the place. Note that these are defined by "config.h", so we need to include that whenever we check `USE_CPPJIEBA`.

rschu1ze commented on src/Interpreters/ITokenExtractor.cpp:
```diff
@@ -15,10 +14,36 @@
 
+std::vector<String> ITokenExtractor::getTokens(const char * data, size_t length) const
+{
+    std::vector<String> tokens;
```
Moved to the base class now. There was some nice .reserve logic but in practice it likely won't make a difference as the first push_back will reserve enough space anyways and we'll do exponential resize.

rschu1ze commented on src/Interpreters/ITokenExtractor.cpp:
```diff
@@ -270,4 +300,46 @@ void SplitTokenExtractor::substringToGinFilter(const char * data, size_t length,
+
+bool ChineseTokenExtractor::nextInString(const char * /*data*/, size_t /*length*/, size_t * __restrict /*pos*/, size_t * __restrict /*token_start*/, size_t * __restrict /*token_length*/) const
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Method nextInStringLike is not supported by ChineseTokenExtractor");
```
There was some implementation here which didn't look finished. Now throws an exception to be on the safe side.

rschu1ze commented on src/Interpreters/ITokenExtractor.cpp:
```diff
@@ -270,4 +300,46 @@ void SplitTokenExtractor::substringToGinFilter(const char * data, size_t length,
+
+bool NoOpTokenExtractor::nextInString(const char * /*data*/, size_t length, size_t * __restrict pos, size_t * __restrict token_start, size_t * __restrict token_length) const
+{
+    if (*pos == 0)
```
Isn't that `*pos != 0`?

rschu1ze commented on src/Interpreters/ITokenExtractor.h:
```diff
@@ -179,8 +194,43 @@ struct SplitTokenExtractor final : public ITokenExtractorHelper<SplitTokenExtrac
+/// Parser doing "no operation". Returns the entire input as a single token.
+struct NoOpTokenExtractor final : public ITokenExtractorHelper<NoOpTokenExtractor>
+{
+    static const char * getName() { return "noop"; }
```
I un-inlined pretty much everything.

rschu1ze commented on src/Interpreters/ITokenExtractor.h:
```diff
@@ -179,8 +194,43 @@ struct SplitTokenExtractor final : public ITokenExtractorHelper<SplitTokenExtrac
+struct NoOpTokenExtractor final : public ITokenExtractorHelper<NoOpTokenExtractor>
+{
+    static const char * getName() { return "noop"; }
+    static const char * getExternalName() { return getName(); }
```
The tokenizer names in namespace `detail` were replaced by `getExternalName()`. We should keep classes and names in the same place.

rschu1ze commented on tests/queries/0_stateless/03403_function_tokens.sql:
```diff
@@ -0,0 +1,71 @@
+-- Tags: no-fasttest
+-- no-fasttest: Chinese tokenization relies on a 3rd party library
+
+SELECT 'Negative tests';
```
Added this section.

rschu1ze commented on tests/queries/0_stateless/03403_function_tokens.sql:
```diff
@@ -0,0 +1,71 @@
+SELECT tokens('', 'noop') AS tokenized;
+SELECT tokens('abc def', 'noop') AS tokenized;
+
+SELECT 'Chinese tokenizer';
```
Deleted the other test so all tests for the new function are in one file.

rschu1ze committed:
Fix style

rschu1ze committed:
Fix docs

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into ahmadov/tokenizer-cn

rschu1ze committed:
Remove Chinese tokenizer for now

rschu1ze committed:
Cosmetics

rschu1ze committed:
Cosmetics

rschu1ze committed:
Update docs

rschu1ze committed:
Fix style

rschu1ze commented:
Stateless tests (msan, 3/4):
- 01275_parallel_mv: #66116

AST fuzzer (amd_debug):
- #79451
